contrib/babelfishpg_common/src/datetime.c

```diff
@@ -797,4 +804,548 @@ datetime_to_numeric(PG_FUNCTION_ARGS)
+					hourdiff = tm2->tm_hour - tm1->tm_hour;
+					minutediff = tm2->tm_min - tm1->tm_min;
+					seconddiff = tm2->tm_sec - tm1->tm_sec;
+					overflow = (overflow || !(int64_multiply_add(daydiff, 24, &hourdiff)));
+					overflow = (overflow || !(int64_multiply_add(hourdiff, 60, &minutediff)));
+					overflow = (overflow || !(int64_multiply_add(minutediff, 60, &seconddiff)));
+					diff = seconddiff;
+					break;
+				case DTK_MILLISEC:
+					daydiff = days_in_date(tm2->tm_mday, tm2->tm_mon, tm2->tm_year) - days_in_date(tm1->tm_mday, tm1->tm_mon, tm1->tm_year);
+					hourdiff = tm2->tm_hour - tm1->tm_hour;
+					minutediff = tm2->tm_min - tm1->tm_min;
+					seconddiff = tm2->tm_sec - tm1->tm_sec;
+					millisecdiff = (fsec2 / 1000) - (fsec1 / 1000);
+					overflow = (overflow || !(int64_multiply_add(daydiff, 24, &hourdiff)));
+					overflow = (overflow || !(int64_multiply_add(hourdiff, 60, &minutediff)));
+					overflow = (overflow || !(int64_multiply_add(minutediff, 60, &seconddiff)));
+					overflow = (overflow || !(int64_multiply_add(seconddiff, 1000, &millisecdiff)));
+					diff = millisecdiff;
+					break;
+				case DTK_MICROSEC:
+					daydiff = days_in_date(tm2->tm_mday, tm2->tm_mon, tm2->tm_year) - days_in_date(tm1->tm_mday, tm1->tm_mon, tm1->tm_year);
+					hourdiff = tm2->tm_hour - tm1->tm_hour;
+					minutediff = tm2->tm_min - tm1->tm_min;
+					seconddiff = tm2->tm_sec - tm1->tm_sec;
+					microsecdiff = fsec2 - fsec1;
+					overflow = (overflow || !(int64_multiply_add(daydiff, 24, &hourdiff)));
+					overflow = (overflow || !(int64_multiply_add(hourdiff, 60, &minutediff)));
+					overflow = (overflow || !(int64_multiply_add(minutediff, 60, &seconddiff)));
+					overflow = (overflow || !(int64_multiply_add(seconddiff, 1000000, &microsecdiff)));
+					diff = microsecdiff;
+					break;
+				case DTK_NANO:
+					daydiff = days_in_date(tm2->tm_mday, tm2->tm_mon, tm2->tm_year) - days_in_date(tm1->tm_mday, tm1->tm_mon, tm1->tm_year);
+					hourdiff = tm2->tm_hour - tm1->tm_hour;
+					minutediff = tm2->tm_min - tm1->tm_min;
+					seconddiff = tm2->tm_sec - tm1->tm_sec;
+					microsecdiff = fsec2 - fsec1;
+					overflow = (overflow || !(int64_multiply_add(daydiff, 24, &hourdiff)));
+					overflow = (overflow || !(int64_multiply_add(hourdiff, 60, &minutediff)));
+					overflow = (overflow || !(int64_multiply_add(minutediff, 60, &seconddiff)));
+					overflow = (overflow || !(int64_multiply_add(seconddiff, 1000000, &microsecdiff)));
+					overflow = (overflow || (pg_mul_s64_overflow(microsecdiff, 1000, &diff)));
+					break;
+				default:
+					validDateDiff = false;
+			}
+		}
+		else {
+			ereport(ERROR,
+				(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
+				 errmsg("timestamp out of range")));
+		}
+	} else {
+		validDateDiff = false;
+	}
+
+	if(!validDateDiff) {
+		elog(ERROR, "The datepart %s is not a recognized datediff option.", lowunits);
```

Comment:
Should add test cases for each error out and make sure transaction behavior is the same as sql server ( when error out, shouldn't abort the current transaction )